Router sticky sessions

pweil- opened the pull request:
@rajatchopra @smarterclayton  - wanted to get an initial review on this config while I work on the peer sync portion.  

Here is some testing: https://gist.github.com/pweil-/c9bfb7f32bad56dd9f34

rajatchopra commented:
Will examine the testing stuff. Looks good otherwise, is at parity with v2 for http.

rajatchopra commented:
Tests look good too.

pweil- commented:
@rajatchopra - I've got the peer pieces set up by passing the service name that the osadm command creates to the router.  It uses that service to capture the endpoints for the router(s) and creates a peer section.  I then modified the start up script to grab the first line out of `/etc/hosts` in the container and force the peer name with the `-L` argument.  But it seems that a container on one minion cannot reach the containers on the other minions to sync the stick tables, which it should be able to do anyway for multiple replicas in a single service (couldn't ping the endpoint from inside the container).  

I've been having some trouble with the sdn on the vagrant setup in general anyway so maybe you can help me out tomorrow and figure out what I'm doing wrong.  For me, it seemed the eth1 interface didn't have an IP address until I ran an ifup, and even then it would periodically go away until I sent some traffic down the pipe.  I started it up with the directions from https://github.com/openshift/openshift-sdn#try-it-out.

```
[vagrant@openshift-master hello-sni]$ osc get pods
POD              IP            CONTAINER(S)   IMAGE(S)                          HOST                            LABELS                                                      STATUS    CREATED      MESSAGE
router-1-g8hut   172.17.0.8                                                     openshift-minion-2/10.245.2.4   deployment=router-1,deploymentconfig=router,router=router   Running   22 minutes   
                               router         openshift/origin-haproxy-router                                                                                               Running   22 minutes   
router-1-m1crt   172.17.0.10                                                    openshift-minion-1/10.245.2.3   deployment=router-1,deploymentconfig=router,router=router   Running   22 minutes   
                               router         openshift/origin-haproxy-router                                                                                               Running   22 minutes   
tls-server       172.17.0.11                                                    openshift-minion-1/10.245.2.3   name=tls-server                                             Running   19 minutes   
                               tls-server     pweil/tls-server                                                                                                              Running   19 minutes   
tls-server2      172.17.0.9                                                     openshift-minion-2/10.245.2.4   name=tls-server                                             Running   15 minutes   
                               tls-server2    pweil/tls-server                                                                                                              Running   15 minutes   
```

From inside the router on minon-1
```
[root@router-1-m1crt /]# ping 172.17.0.11
PING 172.17.0.11 (172.17.0.11) 56(84) bytes of data.
64 bytes from 172.17.0.11: icmp_seq=1 ttl=64 time=0.037 ms
64 bytes from 172.17.0.11: icmp_seq=2 ttl=64 time=0.053 ms
^C
--- 172.17.0.11 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1001ms
rtt min/avg/max/mdev = 0.037/0.045/0.053/0.008 ms
[root@router-1-m1crt /]# ping 172.17.0.9
PING 172.17.0.9 (172.17.0.9) 56(84) bytes of data.
From 172.17.0.10 icmp_seq=1 Destination Host Unreachable
From 172.17.0.10 icmp_seq=2 Destination Host Unreachable
From 172.17.0.10 icmp_seq=3 Destination Host Unreachable
From 172.17.0.10 icmp_seq=4 Destination Host Unreachable
```

rajatchopra commented:
Yes, that setup sounds wrong. 172.17.0.11 as the container's ipaddress suggests the SDN was never setup alright.

pweil- commented:
Ok, got it working enough to test what I needed.  Would still like to sync up with you, I had to do a lot of ifup eth1, restarting of master/node-sdn, and there was also an issue with a protocol error in the kubelet having to do with the /vagrant/openshift.local.volumes that I'm curious if anyone has run in to.


```
[vagrant@openshift-master hello-sni]$ go run client.go -server openshift-minion-1:443
Hello TLS
[vagrant@openshift-master hello-sni]$ go run client.go -server openshift-minion-2:443
Hello TLS

---- first ----
[root@router-1-103sp /]# echo 'show table be_tcp_default-route-passthrough' | socat unix-connect:/var/lib/haproxy/run/haproxy.sock stdio
# table: be_tcp_default-route-passthrough, type: ip, size:1048576, used:1
0x7f940ed44dd4: key=10.245.2.2 use=0 exp=292763 server_id=1

---- second ---- 
[root@router-1-v1aui /]# echo 'show table be_tcp_default-route-passthrough' | socat unix-connect:/var/lib/haproxy/run/haproxy.sock stdio
# table: be_tcp_default-route-passthrough, type: ip, size:1048576, used:1
0x7f561b415de4: key=10.245.2.2 use=0 exp=289992 server_id=1
```

PTAL

pweil- commented:
Testing steps: https://gist.github.com/pweil-/0817e7d37ea8daab6d2f

pweil- commented:
bump for some reviews @rajatchopra @ramr @abhgupta @smarterclayton.  Please take note of how the existing router service is used to generate the peer sync list for the tcp backend stick tables.

ramr commented on images/router/haproxy/conf/haproxy-config.template:
```diff
@@ -155,8 +175,13 @@ backend be_edge_http_{{$cfgIdx}}
+  cookie OPENSHIFT_{{$cfgIdx}}_SERVERID insert indirect nocache httponly
+  {{ else }}
+  cookie OPENSHIFT_EDGE_{{$cfgIdx}}_SERVERID insert indirect nocache httponly secure
+  {{ end }}
```
Hmm, do we want to make the sticky sessions behavior the default? Wouldn't this mean that now all client connections would end up being fixed to the same backend instance? This does have implications if there's a mix of short and long running sessions - we could very well end up with an unbalanced haproxy setup - example one backend instance gets "favored" over all others. 
Also, most applications at scale understand the need to remove the dependency on sticky sessions or use some sort of external scalable session store. Yeah, there'll be a few that require sticky sessions but IMHO, rather than make that the default, it make more sense to have that be the exception.
One option would be to have a field on a ```route``` that says it needs ```session-affinity`` and based on that generate this config.

ramr replied:
Also the ```cookie``` directive takes  ```maxidle/maxlife``` values - we probably want to set that, so that there's some limits on how long the session affinity holds for.

pweil- replied:
@smarterclayton - thoughts on the above?  Perhaps adding `StickySessions bool` to the route object if we want to have these off by default.

smarterclayton replied:
Are individual developers in an enterprise going to be making that decision?  Or the admin?  If we don't *have* to support it per app then we shouldn't until we have a concrete use case for it.

pweil- replied:
Ultimately, no matter where we want to put a flag (on the router command or on a route by route basis) it is up to the underlying implementation to support sticky sessions which is under the administrator's control.  For the OOTB experience, it is always enabled.  We can decide to make that more flexible as we progress since it is additive.  If we make an api change to request sticky sessions we will, of course, be assuming that the underlying router implementation is honoring the request which we have no control over.

smarterclayton replied:
I don't want to add extra flags yet until we know what people will need.

----- Original Message -----
> > @@ -155,8 +175,13 @@ backend be_edge_http_{{$cfgIdx}}
> >    mode http
> >    balance leastconn
> >    timeout check 5000ms
> > +  {{ if (eq $cfg.TLSTermination "") }}
> > +  cookie OPENSHIFT_{{$cfgIdx}}_SERVERID insert indirect nocache httponly
> > +  {{ else }}
> > +  cookie OPENSHIFT_EDGE_{{$cfgIdx}}_SERVERID insert indirect nocache
> > httponly secure
> > +  {{ end }}
> 
> Ultimately, no matter where we want to put a flag (on the router command or
> on a route by route basis) it is up to the underlying implementation to
> support sticky sessions which is under the administrator's control.  For the
> OOTB experience, it is always enabled.  We can decide to make that more
> flexible as we progress since it is additive.  If we make an api change to
> request sticky sessions we will, of course, be assuming that the underlying
> router implementation is honoring the request which we have no control over.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2105/files#r30532371

ramr replied:
So as mentioned earlier, at scale the main cause of concern is that haproxy won't effectively route to free backends. Adding some weights could possibly alleviate this up to a certain extent. Turning this on by default, might a bit weighty (no pun intended).
The question that comes to mind though is - whether the reasoning for doing this by default is to handle the java case of ```JSESSIONID``` or a login session with a generated id - that true/the case here?
I know this is an haproxy-specific solution, but another alternative to not having an extra flag on the route is for the default router (aka haproxy) here to use the ```appsession``` directive that does route similarly - session affinity. That would route based on the presence of a well-known cookie (```JSESSIONID```, ```PHPSESSID``` etc). That said, there's demerits to that as well - using well known names but it might cover I dunno a large percentage (maybe even in the high 90s) of the use cases. Just a thought.

ramr commented on images/router/haproxy/conf/haproxy-config.template:
```diff
@@ -164,8 +189,10 @@ backend be_edge_http_{{$cfgIdx}}
 backend be_tcp_{{$cfgIdx}}
   balance leastconn
   timeout check 5000ms
+  stick-table type ip size 1m expire 5m{{ if (len $.PeerEndpoints) gt 0 }} peers openshift_peers {{ end }}
```
Ignore a part of my above comment re: the imbalance on the haproxy backends. So we have this set for 5 mins, which seems like a short time for session affinity if sticky sessions are desired.  Maybe am not clear here on the exact use case of this - a short note would help. Thx

pweil- replied:
The 5 minutes for stick tables is the max time that an entry will stay without a refresh.  Any activity on the connection should refresh that 5 minute lease on the entry. Five minutes seems like an awful long time to not have any activity (to me) unless, perhaps, some large edi data processing that must pre-process on the client side after establishing the server connection.  Not sure that is really an issue though, I can think of multiple ways to mitigate that risk.

That said, since we have a max size on the table the oldest could be aged out if we had a higher expiry.  I'm not opposed to changing it.

ramr commented on images/router/haproxy/reload-haproxy:
```diff
@@ -12,8 +12,13 @@ if [ -f $pid_file ]; then
+# Assumes that the docker ip is the first entry in the hosts file.
+# This port is non-configurable via the router cmd, it is HAProxy specific.  It should
+# match the port used in the haproxy-config.template
+peer_name=$(awk 'NR==1 {print $1}' /etc/hosts):1937
```
Would it hurt to just bind on ```0.0.0.0:1937``` ?

smarterclayton replied:
Agreed - let's resolve to not do any hacks to get our ip - the pod downward API should report that for us if we need it.

> On May 15, 2015, at 8:09 PM, ramr <notifications@github.com> wrote:
> 
> In images/router/haproxy/reload-haproxy:
> 
> > @@ -12,8 +12,13 @@ if [ -f $pid_file ]; then
> >    old_pid=$(<$pid_file)
> >  fi
> >  
> > +# Assumes that the docker ip is the first entry in the hosts file.
> > +# This port is non-configurable via the router cmd, it is HAProxy specific.  It should
> > +# match the port used in the haproxy-config.template
> > +peer_name=$(awk 'NR==1 {print $1}' /etc/hosts):1937
> Would it hurt to just bind on 0.0.0.0:1937 ?
> 
> —
> Reply to this email directly or view it on GitHub.
>

pweil- replied:
Should've thought of that. :-\

pweil- replied:
hmm, actually, I will need think about this.  This is not the actual binding, it is telling HAProxy what it should consider the local name in lieu of the default host name.  It uses this to match the peer section to know what is 'me' during synchronization (and opens the ip/port that the label matches).  Since I create the peer section based on the endpoints of the service the names are just the ip/ports, hence using the ip/port as the "name" with `-L`.  

For example, rather than ip/port the peer section could've looked like:

```
peers openshift_peers
    peer thisisme 0.0.0.0:1937
    peer notme 1.1.1.1:1937

and on the haproxy command

-L thisisme
```

So perhaps the right solution is using the pod downward api to get the ip address, match that to the port and use a way to label the peer correctly (using ip/port for all notme servers).  Then I can remove the items in the shell script and just run with -L <known static label>

pweil- replied:
@smarterclayton - been trying to sort this out this morning.

1.  The downward api does not support status.podIP
1.  Adding the support for status.podIP locally resulted in an empty env var

It looks like the pod is still in pending state with no ip when the downward api vars values are injected.  So I'm looking for a short term, acceptable solution.  I could inject the client into the plugin and query based on the downward api's name/namespace.
```
E0519 14:37:49.941356   23572 kubelet.go:899] ------------------- makeing field path for status.podIP with pod api.Pod .
...snip...
 Status:api.PodStatus{Phase:"Pending", Conditions:[]api.PodCondition(nil), Message:"", HostIP:"", PodIP:"", ContainerStatuses:[]api.ContainerStatus(nil)}}
```

pweil- replied:
Actually, I'm going to test one more thing before I try and hack something up.  If I can put a static entry for `peer self 0.0.0.0:1937` in the list without it puking due to a "duplicate" with the endpoint ip then we're fine.

smarterclayton replied:
This is fine for now - just leave a TODO: get this from a more consistent source.

----- Original Message -----
> > @@ -12,8 +12,13 @@ if [ -f $pid_file ]; then
> >    old_pid=$(<$pid_file)
> >  fi
> >  
> > +# Assumes that the docker ip is the first entry in the hosts file.
> > +# This port is non-configurable via the router cmd, it is HAProxy
> > specific.  It should
> > +# match the port used in the haproxy-config.template
> > +peer_name=$(awk 'NR==1 {print $1}' /etc/hosts):1937
> 
> @smarterclayton - been trying to sort this out this morning.
> 
> 1.  The downward api does not support status.podIP
> 1.  Adding the support for status.podIP locally resulted in an empty env var
> 
> It looks like the pod is still in pending state with no ip when the downward
> api vars values are injected.  So I'm looking for a short term, acceptable
> solution.  I could inject the client into the plugin and query based on the
> downward api's name/namespace.
> ```
> E0519 14:37:49.941356   23572 kubelet.go:899] ------------------- makeing
> field path for status.podIP with pod api.Pod .
> ...snip...
>  Status:api.PodStatus{Phase:"Pending", Conditions:[]api.PodCondition(nil),
>  Message:"", HostIP:"", PodIP:"",
>  ContainerStatuses:[]api.ContainerStatus(nil)}}
> ```
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2105/files#r30606345

ramr replied:
@pweil-  aah, k that is different if its using it as an id so would using something like ```hostname -s``` work - should be the pod name, AFAICR.

pweil- replied:
@ramr - yeah, the kicker is that the ids need to match across replicas because haproxy uses that to identify the servers during synchronization (so my static self label test failed).   The peer section labels need to be consistent among all peers with the -L flag identifying which peer you are. 

Unfortunately I'm back to the current solution since what I have to work with is the set of endpoints from the router service.

What I need is

```
peers mypeers
    peer haproxy1 192.168.0.1:1024
    peer haproxy2 192.168.0.2:1024
    peer haproxy3 10.2.0.1:1024
```

where each haproxy instance can be started with their own `-L <my label>` that matches one of those items in peers.

Suggestion box open.  Requirements 

1.  must be able to identify the label for the replica in the restart script for -L
1.  must be able to generate the peer section on all replicas based on the set of endpoints (with a label that matches item 1)

/cc @rajatchopra

pweil- replied:
> @pweil- aah, k that is different if its using it as an id so would using something like hostname -s work - should be the pod name, AFAICR.

To be clear why this doesn't work and why injecting the client wouldn't work: I cannot get the host name of the other replicas to generate the peer section with identical entries.  This only identifies the current replica.

I have the service name.  I could get the selector from the service, query the apiserver to get the pods that match the selector, and build endpoints and labels based on that

```
[vagrant@openshift-master hello-sni]$ osc get services
NAME            LABELS                                    SELECTOR          IP(S)            PORT(S)
kubernetes      component=apiserver,provider=kubernetes   <none>            172.30.0.2       443/TCP
kubernetes-ro   component=apiserver,provider=kubernetes   <none>            172.30.0.1       80/TCP
router          router=router                             router=router     172.30.208.190   80/TCP
tls-service     <none>                                    name=tls-server   172.30.31.12     2999/tcp
[vagrant@openshift-master hello-sni]$ osc get pods -l "router=router"
POD              IP          CONTAINER(S)   IMAGE(S)                          HOST                            LABELS                                                      STATUS    CREATED      MESSAGE
router-1-1hau2   10.1.0.18                                                    openshift-minion-1/10.245.2.3   deployment=router-1,deploymentconfig=router,router=router   Running   25 minutes   
                             router         openshift/origin-haproxy-router                                                                                               Running   25 minutes   
router-1-56kxa   10.1.1.14                                                    openshift-minion-2/10.245.2.4   deployment=router-1,deploymentconfig=router,router=router   Running   25 minutes   
                             router         openshift/origin-haproxy-router                                                                                               Running   25 minutes   
[vagrant@openshift-master hello-sni]$ 
```
The host name is available in an env var that the router could use for the -L

```
[vagrant@openshift-minion-1 hello-sni]$ docker exec 36c6 env
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=router-1-1hau2
```

pweil- replied:
Hold the presses, I may have a solution!

smarterclayton replied:
Talk to mfojtik about DNS if you have to.

----- Original Message -----
> > @@ -12,8 +12,13 @@ if [ -f $pid_file ]; then
> >    old_pid=$(<$pid_file)
> >  fi
> >  
> > +# Assumes that the docker ip is the first entry in the hosts file.
> > +# This port is non-configurable via the router cmd, it is HAProxy
> > specific.  It should
> > +# match the port used in the haproxy-config.template
> > +peer_name=$(awk 'NR==1 {print $1}' /etc/hosts):1937
> 
> > @pweil- aah, k that is different if its using it as an id so would using
> > something like hostname -s work - should be the pod name, AFAICR.
> 
> To be clear why this doesn't work and why injecting the client wouldn't work:
> I cannot get the host name of the other replicas to generate the peer
> section with identical entries.  This only identifies the current replica.
> 
> I have the service name.  I could get the selector from the service, query
> the apiserver to get the pods that match the selector, and build endpoints
> and labels based on that
> 
> ```
> [vagrant@openshift-master hello-sni]$ osc get services
> NAME            LABELS                                    SELECTOR
> IP(S)            PORT(S)
> kubernetes      component=apiserver,provider=kubernetes   <none>
> 172.30.0.2       443/TCP
> kubernetes-ro   component=apiserver,provider=kubernetes   <none>
> 172.30.0.1       80/TCP
> router          router=router                             router=router
> 172.30.208.190   80/TCP
> tls-service     <none>                                    name=tls-server
> 172.30.31.12     2999/tcp
> [vagrant@openshift-master hello-sni]$ osc get pods -l "router=router"
> POD              IP          CONTAINER(S)   IMAGE(S)
> HOST                            LABELS
> STATUS    CREATED
> MESSAGE
> router-1-1hau2   10.1.0.18
> openshift-minion-1/10.245.2.3
> deployment=router-1,deploymentconfig=router,router=router   Running   25
> minutes
>                              router         openshift/origin-haproxy-router
>                              Running
>                              25 minutes
> router-1-56kxa   10.1.1.14
> openshift-minion-2/10.245.2.4
> deployment=router-1,deploymentconfig=router,router=router   Running   25
> minutes
>                              router         openshift/origin-haproxy-router
>                              Running
>                              25 minutes
> [vagrant@openshift-master hello-sni]$
> ```
> The host name is available in an env var that the router could use for the -L
> 
> ```
> [vagrant@openshift-minion-1 hello-sni]$ docker exec 36c6 env
> PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
> HOSTNAME=router-1-1hau2
> ```
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2105/files#r30635296

ramr commented on plugins/router/template/router.go:
```diff
@@ -76,6 +86,8 @@ func newTemplateRouter(templates map[string]*template.Template, reloadScriptPath
 		defaultCertificate:     defaultCertificate,
 		defaultCertificatePath: "",
+		peerEndpointsKey:       peerEndpointsKey,
+		peerEndpoints:          map[string]Endpoint{},
```
Is ```peerEndpoints``` populated somewhere else in this file? The only place I could find it was in ```DeleteEndpoints```.

ramr replied:
Ignore this - found it in ```AddEndpoints``` as well.

pweil- commented:
@ramr - behold!  The endpoint has a target object reference which has a name.  I introduced a new field on the internal endpoint object populated with either the name or ip (since the ref can be nil).  I then used the `HOSTNAME` env var.  There is a possibility of a mismatch if we're using `HOSTNAME` in the reload script but the endpoints don't have the name set.  

@smarterclayton can you comment on the reliability of this?

End result is a config that looks like 
```
peers openshift_peers
    peer router-1-bz89l 10.1.0.26:1937 
    peer router-1-rzqpa 10.1.1.18:1937
```

A reload script with 
```
# Must match an endpoint target name that is used in the peer section
peer_name=$HOSTNAME
```

And a successful test of
```
[root@router-1-bz89l /]# echo 'show table be_tcp_default-route-passthrough' | socat unix-connect:/var/lib/haproxy/run/haproxy.sock stdio
# table: be_tcp_default-route-passthrough, type: ip, size:1048576, used:2
0x7fd607719e74: key=10.1.0.1 use=0 exp=294231 server_id=2
0x7fd6077116a4: key=10.245.2.2 use=0 exp=276364 server_id=1

[root@router-1-rzqpa /]# echo 'show table be_tcp_default-route-passthrough' | socat unix-connect:/var/lib/haproxy/run/haproxy.sock stdio
# table: be_tcp_default-route-passthrough, type: ip, size:1048576, used:2
0x7f02599f1e84: key=10.1.0.1 use=0 exp=291465 server_id=2
0x7f02599f1ae4: key=10.245.2.2 use=0 exp=277130 server_id=1
```

Squashed, @rajatchopra @ramr - can you take another (hopefully final) pass?

pweil- commented:
[test]

ramr commented:
Only concern i have is the one above - enabling sticky-sessions by default. Other than that LGTM.

smarterclayton commented:
Is sticky-sessions that over 50% of admins would expect to have by default?

----- Original Message -----
> Only concern i have is the one above - enabling sticky-sessions by default.
> Other than that LGTM.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2105#issuecomment-103743193

ramr commented:
Well, really depends on the services (the apps if you want it call it that) being hosted? So I dunno, I probably wouldn't classify it as an admin requirement - more than likely an app requirement.
Examples - there's a tomcat/jboss app that uses JSESSIONID to save state locally (though in all fairness there could/should be a backing store/cache - ala memcache/dynamo where that could be persisted and thus allows any instance to service the request) or a login screen which generates a locally stored short-term captcha that needs to be matched on submittal. 

And on the other end of the spectrum, most newer style REST-style microservices or api services if designed well would have a stateless constraint (think hypertext as engine of state) and so really {sh,w}ouldn't have such a requirement. If any state needs to be carried over ala authorization, those services would use http primitives - in this example HTTP ```Authorization``` headers and possibly something like ```OAUTH[2]```.

smarterclayton commented:
50% of our users and admins today.  @brenton?

----- Original Message -----
> Well, really depends on the services (the apps if you want it call it that)
> being hosted? So I dunno, I probably wouldn't classify it as an admin
> requirement - more than likely an app requirement.
> Examples - there's a tomcat/jboss app that uses JSESSIONID to save state
> locally (though in all fairness there could/should be a backing store/cache
> - ala memcache/dynamo where that could be persisted and thus allows any
> instance to service the request) or a login screen which generates a locally
> stored short-term captcha that needs to be matched on submittal.
> 
> And on the other end of the spectrum, most newer style REST-style
> microservices or api services if designed well would have a stateless
> constraint (think hypertext as engine of state) and so really {sh,w}ouldn't
> have such a requirement. If any state needs to be carried over ala
> authorization, those services would use http primitives - in this example
> HTTP ```Authorization``` headers and possibly something like ```OAUTH[2]```.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2105#issuecomment-103972656

pweil- commented:
@brenton - whatcha think based on our v2 knowledge?

smarterclayton commented:
Let's just do source ip by default, and let sticky be enabled.  I'd prefer stateless routers as the default anyway - I'm concerned about the risk.  We'll document the different modes and trade offs for customers

> On May 22, 2015, at 11:18 AM, Paul Weil <notifications@github.com> wrote:
> 
> @brenton - whatcha think based on our v2 knowledge?
> 
> —
> Reply to this email directly or view it on GitHub.
>

pweil- commented:
Discussed in IRC.  Here's the plan

1.  test locally to ensure `balance source` is consistent across replicas
2.  change to `balance source` by default for all backends
3.  retain the code that identifies and labels endpoints by pod.Name and supplies a peer endpoint table
4.  we provide default stickiness based on IP and allow anyone who wants to customize the template to use cookies of their choosing, stick-tables, or cookies with stick-tables by providing all the necessary data to the template router as shown in this PR.

Known issues: users behind a proxy (corporate or otherwise) that result in the same IP address will stick to the same backend server by default.

pweil- commented:
re[test]

pweil- commented:
@rajatchopra @ramr @smarterclayton 

This commit now contains

1.  The changes for finding the service created by the osadm command which can be used to identify peers (to enable peer sections if someone wants to use them)
1.  Change of the EndpointsTable from a map to slice
1.  Sorting of aforementioned slice by ID (which is ip:port)
1.  Balance by source in the template

In the handle endpoints method I've also added a check to see if the endpoints actually changed (the deep equal) which I intend on building on (in a subsequent PR) for https://github.com/openshift/origin/issues/1592 to provide feedback to the plugin so it can tell if a commit is necessary or not.

@bmeng - for testing, it is no longer required to supply cookies for http routes or check stick tables for tcp routes, it should all just work via the hashing, even with multiple routers.

ramr commented:
LGTM - might also want @rajatchopra to take another look.

pweil- commented:
bump, @rajatchopra @abhgupta thoughts, suggestions, concerns?

/cc @brenton - the more the merrier!

pweil- commented:
@abhgupta @rajatchopra  @smarterclayton - ptal

smarterclayton commented on pkg/cmd/experimental/router/router.go:
```diff
@@ -301,3 +316,19 @@ func RunCmdRouter(f *clientcmd.Factory, cmd *cobra.Command, out io.Writer, cfg *
+// getServiceNameFromObjects operates on the assumption that there will only be a single service
+// in the list of objects that it can use to give back a namespace and name to be used in the
+// router's environment if it needs to reference its peers
+func getServiceNameFromObjects(objects []runtime.Object) (string, error) {
```
Why is this necessary?  The service name should be the same as the deployment config name always.

pweil- replied:
that's true, will remove.  @pweil- doing it the hard way since 2014.

smarterclayton commented on plugins/router/template/router.go:
```diff
@@ -44,19 +46,29 @@ type templateRouter struct {
-func newTemplateRouter(templates map[string]*template.Template, reloadScriptPath, defaultCertificate string) (*templateRouter, error) {
+func newTemplateRouter(templates map[string]*template.Template, reloadScriptPath, defaultCertificate string, peerEndpointsKey string) (*templateRouter, error) {
 	glog.Infof("Creating a new template router")
+	glog.Infof("Router will use %s keyed endpoints to identify peers", peerEndpointsKey)
```
Message isn't obvious what it means.

smarterclayton commented on plugins/router/template/router.go:
```diff
@@ -326,16 +345,19 @@ func (r *templateRouter) RemoveRoute(id string, route *routeapi.Route) {
-			newEndpoint := Endpoint{ep.ID, ep.IP, ep.Port}
-			frontend.EndpointTable[ep.ID] = newEndpoint
+	//sort the endpoints
+	eps := EndpointSlice(endpoints)
```
Why are the endpoints in a different order than what comes out of the endpoints object?

smarterclayton commented on plugins/router/template/router.go:
```diff
@@ -326,16 +345,19 @@ func (r *templateRouter) RemoveRoute(id string, route *routeapi.Route) {
+	eps := EndpointSlice(endpoints)
+	sort.Sort(eps)
+	//only make the change if there is a difference
+	if !reflect.DeepEqual(frontend.EndpointTable, endpoints) {
```
You should return if they are equal, not nest the complex case

pweil- commented:
1.  removed the crazy service lookup
1.  removed the EndpointSlice - not necessary since we're not sorting.  Just use the endpoints as a slice now rather than the map as it was before
1.  updated the log message
1.  changed the comparison logic

danmcp commented:
@pweil- What's the next step here?

pweil- commented:
@danmcp - it is complete and waiting on a lgtm.  Needs a rebase now though, I'll take care of that this morning.

danmcp commented:
@rajatchopra @ramr @smarterclayton Review please.

smarterclayton commented:
LGTM [merge]
